File: leetcode/medium/PermutationInString.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        """
        Checks if a permutation of s1 exists as a substring in s2.

        This solution uses a sliding window approach with a fixed-size window
        equal to the length of s1. Instead of comparing full frequency maps for
        each window, it uses a single array `freq_diff` to track the difference
        in character frequencies between s1 and the current window in s2.

        The core idea is to maintain a `matches` count, which tracks how many
        characters have a frequency difference of zero. When `matches` equals 26,
        it means the character counts in s1 and the current window are identical,
        and thus a permutation has been found. This allows for O(1) updates
        per window slide.

        The overall time complexity is O(len(s2)), and space complexity is O(1).

        Args:
            s1: The pattern string.
            s2: The string to search within.

        Returns:
            True if a permutation of s1 is a substring of s2, False otherwise.
        """
        if len(s1) > len(s2):
            return False

        # freq_diff will store the difference in frequencies.
        # A value of 0 means the frequency is the same in s1 and the window.
        freq_diff = [0] * 26
        s1_len = len(s1)

        # Initialize the first window and the frequency difference map.
        for i in range(s1_len):
            freq_diff[ord(s1[i]) - ord('a')] += 1
            freq_diff[ord(s2[i]) - ord('a')] -= 1

        # Count how many character frequencies are already perfectly matched.
        matches = 0
        for freq in freq_diff:
            if freq == 0:
                matches += 1
        
        if matches == 26:
            return True

        # Slide the window across s2.
        for i in range(s1_len, len(s2)):
            # Character entering the window from the right.
            right_char_idx = ord(s2[i]) - ord('a')
            # Character leaving the window from the left.
            left_char_idx = ord(s2[i - s1_len]) - ord('a')

            # --- Update for the entering character (right_char) ---
            # This character is now in our window, so we decrement its count
            # from the perspective of the s1 map.
            
            # If its frequency diff was 0, we are losing a match.
            if freq_diff[right_char_idx] == 0:
                matches -= 1
            freq_diff[right_char_idx] -= 1
            # If its frequency diff is now 0, we have gained a match.
            if freq_diff[right_char_idx] == 0:
                matches += 1

            # --- Update for the leaving character (left_char) ---
            # This character is leaving our window, so we add its count back
            # from the perspective of the s1 map.

            # If its frequency diff was 0, we are losing a match.
            if freq_diff[left_char_idx] == 0:
                matches -= 1
            freq_diff[left_char_idx] += 1
            # If its frequency diff is now 0, we have gained a match.
            if freq_diff[left_char_idx] == 0:
                matches += 1

            # Check for a full match after the updates.
            if matches == 26:
                return True

        return False

File: leetcode/medium/PermutationInString.java
class Solution {
    /**
     * Checks if s2 contains a permutation of s1.
     * Uses a Sliding Window approach with an optimized frequency match counter.
     *
     * @param s1 The substring to search for (permutation).
     * @param s2 The main string to search within.
     * @return true if a permutation of s1 exists in s2, false otherwise.
     */
    public boolean checkInclusion(String s1, String s2) {
        int len1 = s1.length();
        int len2 = s2.length();

        if (len1 > len2) {
            return false;
        }

        // Frequency array stores the count difference between s1 and the current window in s2.
        // Positive value: s1 has more of this char.
        // Negative value: current window has more of this char.
        // Zero: counts match perfectly.
        int[] countDiff = new int[26];

        // Initialize the window with the first len1 characters
        for (int i = 0; i < len1; i++) {
            countDiff[s1.charAt(i) - 'a']++;
            countDiff[s2.charAt(i) - 'a']--;
        }

        // Count how many characters currently have a perfect match (diff is 0)
        int matches = 0;
        for (int i = 0; i < 26; i++) {
            if (countDiff[i] == 0) {
                matches++;
            }
        }
        
        // If all 26 characters match immediately, return true
        if (matches == 26) {
            return true;
        }

        // Slide the window across s2
        // i represents the index of the character ENTERING the window
        for (int i = len1; i < len2; i++) {
            int rightCharIdx = s2.charAt(i) - 'a';        // New char entering window
            int leftCharIdx = s2.charAt(i - len1) - 'a';  // Old char leaving window

            // Optimization: If entering and leaving chars are the same, the balance doesn't change.
            if (rightCharIdx == leftCharIdx) {
                continue;
            }

            // Update for the entering character (s2 adds to the window, so we decrement the balance)
            if (countDiff[rightCharIdx] == 0) {
                matches--; // It was balanced, now it's not
            }
            countDiff[rightCharIdx]--;
            if (countDiff[rightCharIdx] == 0) {
                matches++; // It is now balanced
            }

            // Update for the leaving character (s2 removes from window, so we increment the balance)
            if (countDiff[leftCharIdx] == 0) {
                matches--; // It was balanced, now it's not
            }
            countDiff[leftCharIdx]++;
            if (countDiff[leftCharIdx] == 0) {
                matches++; // It is now balanced
            }

            if (matches == 26) {
                return true;
            }
        }

        return false;
    }
}